### ISSUE TITLE
Reduce initial delay of Stargate readiness probe

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra#654]
**Is your feature request related to a problem? Please describe.**
The Stargate deployment readiness probe has an initial delay of 4 minutes. This means that the kubelet will wait 4 minutes before checking that the pods are ready. This really slows down startup down. It also means that clients from outside the cluster will not be able to access Stargate during this time.

**Describe the solution you'd like**
The Stargate deployment already uses an initContainer to wait for at least one Cassandra pod to become ready. It needs to be updated to wait until the whole C* cluster is ready (see #612 for details). Given the delay from the initContainer we should be able to safely reduce the initial delay substantially, maybe to 10 or 15 seconds. When #612 is fixed, we should be able to further reduce it. 

We are currently **not** setting the failure threshold. We can set that in conjunction with the reduced initial delay.

The initial delay of the liveness probe is also set to 4 minutes. That can probably be safely reduced quite a bit as well, but I don't think it will impact or delay when the pods become ready.

As part of the work for this it would be nice to see some analysis on what the impact is for using different different settings for the readiness probe as well as the liveness probe.

**Describe alternatives you've considered**

**Additional context**
I want to point out that when we implement controllers for k8ssandra and Stargate, we won't need the initContainer any more. The controller can periodically check on the CassandraDatacenter. When it is ready, it can then deploy Stargate. We already do this in reaper-operator for Reaper.

**Comment by adutra** [comment]
I did some investigation today. I think we can split the changes in 2 areas:

### 1. Changes to the init container

I explored a few ideas:

- Interrogate the K8s API as explained [here](https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/): unfortunately, I ran into security issues as the pod's service account is not allowed to gather info about pods or other resources.
- Enhance the current `nslookup` hack. Instead of waiting for at least one A record, wait until we have N A records, N being the size of datacenter 0. 

I will use the last method for the time being.

### 2. Changes to the Stargate health probes

I verified that the http probes are only available when Stargate's healthchecker component is activated, and the Jetty server on 8084 starts up. Which, on my laptop, happens ~50 seconds after Stargate is started. Before that of course, a liveness check would fail with "Connection refused".

Initially, the probes return DOWN (503) / NOT READY (503). The probes only return UP (200) / READY (200) when Stargate is entirely started, which on my laptop happens roughly at ~60 seconds. 

A few remarks:

1. I don't think we can set the probes `initialDelaySeconds` to less than, say, ~60 seconds; i tried, and the pod keeps being killed. To be on the safe side, I would suggest `initialDelaySeconds: 120` for both probes. 
2. I find it a bit odd that the liveness probe returns DOWN (503) initially; if should return 200 as soon as it is functioning IMO. By returning a 503 status code before Stargate is ready, it delays the moment where K8s could consider the Pod as Running and acts more like a readiness probe than a liveness one.

**Comment by jsanda** [comment]
@adutra thanks for the detailed write up!

The `nslookup` hack sounds good for now. We can solve this better when we implement a controller for Stargate.

What you describe about the liveness probe does not surprise me. My local Stargate repo is not up to date but here is what I have for the liveness check in `CheckResources.java`:

```java
 @GET
  @Path("/liveness")
  public Response checkLiveness() {
    // Sufficient to just return a 200 OK since liveness just means "app is up and doesn't need
    // restarted"
    HealthCheck bundleHealthCheck = healthCheckRegistry.getHealthCheck(BUNDLES_CHECK_NAME);
    if (bundleHealthCheck == null || !bundleHealthCheck.execute().isHealthy()) {
      return Response.status(Response.Status.SERVICE_UNAVAILABLE).entity("DOWN").build();
    }

    HealthCheck schemaHealthCheck = healthCheckRegistry.getHealthCheck(SCHEMA_CHECK_NAME);
    if (schemaHealthCheck == null || !schemaHealthCheck.execute().isHealthy()) {
      return Response.status(Response.Status.SERVICE_UNAVAILABLE).entity("DOWN").build();
    }

    return Response.status(Response.Status.OK).entity("UP").build();
  }
```
I do not think there should be any sort of schema check for the liveness probe. This might be something we want to raise upstream with the Stargate project.

**Comment by adutra** [comment]
@jsanda thanks for the pointers above. Searching for issues in Stargate, I came across https://github.com/stargate/stargate/issues/636, where the schema check was added to the liveness probe. I get the rationale, but I struggle to understand why restarting the pod would solve a schema disagreement issue. I left a comment there, but I guess we'll have to live with it for now.

**Comment by dimas-b** [comment]
Just had another idea. We could probably make the HTTP endpoint for liveness and readiness probes available very soon after Stargate startup.

Currently this endpoint follows the OSGi boostrapping process with dependency links to other services, which delays it availability as an HTTP service. We could make those dependencies "lazy" and make the service start first. It'd take some fiddling with OSGi configuration and would be a bit unnatural in the current Stargate codebase, but it's doable if it helps k8s-based applications. WDYT?

**Comment by jsanda** [comment]
Hey @dimas-b 

Can you provide a brief summary of what the liveness and readiness probes in Stargate do? 

I ask because there is a distinct difference between the two in Kubernetes. In Kubernetes if a liveness probe fails the container is restarted. For this reason in general a liveness probe should be a simple, quick check to indicate that the application is responsive. The readiness probe is different. Clients access Stargate through a Kubernetes service. That service keeps track of the Stargate's IP address to handle routing requests. If the readiness probe fails, the ip address (or Endpoint in terms of Kubernetes types) is removed from the service.

**Comment by dimas-b** [comment]
Currently the liveness probe in SG will fail in two cases: 1) some modules (e.g. GraphQL) fail to initialize; 2) Stargate gives up on achieving schema agreement with storage nodes. In either case, the whole Stargate process is probably not very usable for service client requests.

The readiness probe validates that `CQL`, `REST` and `GraphQL` API are available, that the JVM does not have thread deadlocks, that the internal `DataStore` is functional and (if configured, ON in Astra) that Stargate is able to read from storage nodes.

With respect to your message, @jsanda, it seems to be that the deadlock check should be moved to the "liveness" probe instead. WDYT?

Please note those that the failure modes I indicated above apply only when the health checker HTTP endpoint is up. Currently, it comes up pretty late during the SG startup process. Before that, both "readiness" and "liveness" URLs will get 404 responses, if I'm not mistaken. However the health check URLs can be made available earlier. That's what I meant in my previous message.

**Comment by dougwettlaufer** [comment]
>  However the health check URLs can be made available earlier. 

I'm not sure there's a benefit in that since the probe will be looking for a 200 response. If it's available earlier its going to be returning 503s instead of 404s.

**Comment by dimas-b** [comment]
I agree. With @ivansenic 's pointer to the "startup" probe, making the liveness endpoint available earlier does not seem necessary.

I think we should still move the deadlock check from the "readiness" to the "liveness" probe in Stargate.

**Comment by jsanda** [comment]
>The only problem can be the Kubernetes version we need to support, but I see it exsits in v1.17 and I hope requirements are not below that.

K8ssandra still supports 1.16.

**Comment by jsanda** [comment]
>Currently the liveness probe in SG will fail in two cases: 1) some modules (e.g. GraphQL) fail to initialize; 2) Stargate gives up on achieving schema agreement with storage nodes. In either case, the whole Stargate process is probably not very usable from the client perspective.

@dimas-b does the liveness probe fail while modules are initializing? Or can it succeed during initialization and then will fail if module initialization fails?

I hope my question makes sense 😄

**Comment by dimas-b** [comment]
I believe it is hypothetically possible for the liveness probe to report transient failures during Stargate startup. 

In practice, though, I believe the `liveness` endpoint is the last one to become available over HTTP, so I believe all other module are already in their final state (solid "available", or solid "error") by the time Stargate starts responding to liveness probes.

In any case I assume we will fix this under stargate/stargate#934
